batcher/aligned-batcher/src/eth/utils.rs

```diff
@@ -1,12 +1,137 @@
 use std::str::FromStr;
+use std::sync::Arc;
 
+use crate::config::ECDSAConfig;
+use crate::retry::batcher_retryables::{get_current_nonce_retryable, get_gas_price_retryable};
+use crate::retry::retry_function;
+use crate::types::errors::BatcherError;
+use aligned_sdk::core::constants::{
+    DEFAULT_BACKOFF_FACTOR, DEFAULT_MAX_RETRIES, DEFAULT_MIN_RETRY_DELAY,
+    OVERRIDE_GAS_PRICE_MULTIPLIER, PERCENTAGE_DIVIDER,
+};
+use ethers::prelude::*;
 use ethers::providers::{Http, Provider};
+use log::error;
 
-pub(crate) const GAS_MULTIPLIER: f64 = 1.125; // Multiplier for the gas price for gas escalator
-pub(crate) const GAS_ESCALATOR_INTERVAL: u64 = 12; // Time in seconds between gas escalations
+use super::payment_service::SignerMiddlewareT;
 
 pub fn get_provider(eth_rpc_url: String) -> Result<Provider<Http>, anyhow::Error> {
     let provider = Http::from_str(eth_rpc_url.as_str())
         .map_err(|e| anyhow::Error::msg(format!("Failed to create provider: {}", e)))?;
     Ok(Provider::new(provider))
 }
+
+pub async fn get_batcher_signer(
+    provider: Provider<Http>,
+    ecdsa_config: ECDSAConfig,
+) -> Result<Arc<SignerMiddlewareT>, anyhow::Error> {
+    let chain_id = provider.get_chainid().await?;
+
+    // get private key from keystore
+    let wallet = Wallet::decrypt_keystore(
+        &ecdsa_config.private_key_store_path,
+        &ecdsa_config.private_key_store_password,
+    )?
+    .with_chain_id(chain_id.as_u64());
+
+    let signer = Arc::new(SignerMiddleware::new(provider, wallet));
+    Ok(signer)
+}
+
+/// Calculates an increased gas price for retrying a transaction override.
+/// The gas price rises with each retry by applying a multiplier based on the iteration count.
+pub fn get_bumped_gas_price(
+    previous_gas_price: U256,
+    current_gas_price: U256,
+    iteration: usize,
+) -> U256 {
+    let override_gas_multiplier = U256::from(OVERRIDE_GAS_PRICE_MULTIPLIER) + (5 * iteration);
+    let bumped_previous_gas_price =
+        previous_gas_price * override_gas_multiplier / U256::from(PERCENTAGE_DIVIDER);
+
+    let bumped_current_gas_price =
+        current_gas_price * override_gas_multiplier / U256::from(PERCENTAGE_DIVIDER);
+    bumped_current_gas_price.max(bumped_previous_gas_price)
+}
```

Comment:
Why are we calculating two gas prices and choosing the max?

Reply:
want to cover the case where the `gas_price` decreases and the `previous_gas_price` sent is higher than the current one, to avoid sending a transaction with a lower gas_price than the previous one.